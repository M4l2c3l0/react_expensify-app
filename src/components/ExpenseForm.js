import React from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';

const date = moment();
console.log(date.format("Do MMM, YYYY"));

export default class Expenseform extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : '',
      notes: props.expense ? props.expense.notes : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      focused: false,
      error: ''
    };
  };
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;

    if(amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };
  onDateChange = (createdAt) => {
    if(createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ focused: focused }));
  };
  onNotesChange = (e) => {
    const notes = e.target.value;
    this.setState(() => ({ notes }));
  };
  onSubmit = (e) => {
    e.preventDefault();

    if(!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please enter all required information.' }));
    } else {
      this.setState(() => ({ error: '' }))
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.notes
      });
    }
  }
  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            value={this.state.description}
            onChange={this.onDescriptionChange}
            autoFocus 
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.focused}
            onFocusChange={this.onFocusChange}
            id="singleDayCalendar"
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Notes (optional)..."
            value={this.state.notes}
            onChange={this.onNotesChange}
          ></textarea>
          <button>Add</button>
        </form>
      </div>
    );
  };
};