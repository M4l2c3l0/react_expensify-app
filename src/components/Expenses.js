import React from 'react';
import { connect } from 'react-redux';
import Expense from './Expense';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
  <div>
    {
      props.expenses.length === 0 ? (
        <p>Add your first expense!</p>
      ) : (
        props.expenses.map((expense) => {
          <Expense key={expense.id} {...expense} />;
        })
      )
    }
    {props.expenses.map((expense) => {
      return <Expense {...expense} key={expense.id} />
    })}
  </div>
);

const mapStateToProps = (state) => ({
  expenses: selectExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpenseList);