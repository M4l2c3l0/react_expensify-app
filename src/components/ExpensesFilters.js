import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';

class ExpensesFilters extends React.Component {
  state = {
    focused: null
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.dispatch(setStartDate(startDate));
    this.props.dispatch(setEndDate(endDate));
  };
  onFocusChange = (focused) => {
    this.setState(() => ({ focused }));
  };
  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={(e) => {
            this.props.dispatch(setTextFilter(e.target.value))
          }}
        />
        <select
          value={this.props.filters.sortBy}
          onChange={(e) => {
            this.props.dispatch(e.target.value === 'amount' ? sortByAmount() : sortByDate());
          }}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.focused}
          onFocusChange={this.onFocusChange}
          showClearDates={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
          id="dateRangeCalendar"
        />
      </div>
    );
  }
};

const mapStateToProps = (state) => ({
  filters: state.filters
});

export default connect(mapStateToProps)(ExpensesFilters);