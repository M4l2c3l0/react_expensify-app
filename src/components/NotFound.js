import React from 'react';
import { Link } from 'react-router-dom';

const NotFound = () => (
  <div>
    404 Not Found<br />
    <Link to="/">Go to Dashboard</Link>
  </div>
);

export default NotFound;