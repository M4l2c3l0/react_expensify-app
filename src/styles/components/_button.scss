.button {
  background: $blue;
  border: none;
  color: #fff;
  display: inline-block;
  font-size: $large-font;
  font-weight: 300;
  line-height: 1;
  padding: $s-size;
  text-decoration: none;
}

.button--link {
  background: none;
}

.button--secondary {
  background: transparent;
  border: 1px solid #ec5252;
  color: #ec5252;
  transition: background .3s ease;
  &:hover {
    background: #ec5252;
    color: #fff;
  }
}