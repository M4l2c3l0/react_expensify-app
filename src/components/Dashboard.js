import React from 'react';
import Expenses from './Expenses';
import ExpensesFilters from './ExpensesFilters';

const Dashboard = () => (
  <div>
    <ExpensesFilters />
    <Expenses />
  </div>
);

export default Dashboard;