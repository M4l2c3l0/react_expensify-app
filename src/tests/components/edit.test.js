import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import { Edit } from '../../components/Edit';

let editExpense, removeExpense, history, wrapper;

beforeEach(() => {
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = jest.fn();
  wrapper = shallow(
    <Edit
      editExpense={editExpense}
      removeExpense={removeExpense}
      history={history}
      expense={expenses[2]}
    />
  )
});

test('should render Edit', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle editExpense', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[2]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(editExpense).toHaveBeenLastCalledWith(expenses[2].id, expenses[2]);
});

test('should handle removeExpense', () => {
  wrapper.find('button').simulate('click');
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(removeExpense).toHaveBeenLastCalledWith({
    id: expenses[2].id
  });
});