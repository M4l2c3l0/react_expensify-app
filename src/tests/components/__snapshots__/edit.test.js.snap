// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Edit 1`] = `
<div>
  <Expenseform
    expense={
      Object {
        "amount": 120000,
        "createdAt": 345600000,
        "description": "Credit Card",
        "id": "3",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
